Stage 3

irmo12 opened the pull request.

irmo12 committed:
node error

irmo12 committed:
init stage 3

irmo12 committed:
init newsApi

irmo12 committed:
searching init

irmo12 committed:
searching init

irmo12 committed:
preparing results block

irmo12 committed:
maleware node problem commit

irmo12 committed:
some logical fixes in searchform,
newsApi WIP,

irmo12 committed:
response calibrations

irmo12 committed:
partial news card calibrations

irmo12 committed:
add keyword to articles.
init button in card, next up saved state and prop drill

irmo12 committed:
newscard enhancements,
prep app to handle saving and deleting

irmo12 committed:
enhance card,
break saveordel func,
facepalm

irmo12 committed:
almost done, unsave?
preloader broken,

irmo12 committed:
rev 0

irmo12 committed:
protected routing break

irmo12 committed:
real route protection remains

gennady-bars reviewed (commented):
Dear Omri , (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project:



* You send requests to `localhost` but for review you need to send them to your deployed backend https://disk.yandex.ru/i/4PkORKi8bkCBkQ  . So, I can't check the logics of the app


I've noticed some mistakes:


* You need to deploy the app and show the link in `Readme`
* In ` README.md` you need to add some info about your project: technologies, how to download, how to start, how to deploy the app. And you need to add the link to your **deployed app**.
* When I search anything there is a `key` error https://disk.yandex.ru/i/Zr4ohBTaFQQziQ

irmo12 committed:
dep

irmo12 committed:
dep

irmo12 committed:
dep

irmo12 committed:
thank Jeshua

irmo12 committed:
readme

gennady-bars commented on news-explorer/src/components/PopupWithForm/PopupWithForm.js:
```diff
@@ -8,10 +8,11 @@ export default function PopupWithForm({
+
 
   return (
     <div className={`popup ${isOpen ? 'popup_active' : ''}`} id={name}>
```
**Could be improved**


If it’s interesting for you here is a way to make a special wrapper for any `popup` with cross button, `Escape` and `overlay` listeners inside. You should create component `Popup`, store it in `Popup.js`. 

```jsx
import { useEffect } from "react";

const Popup = ({ isOpen, name, onClose, children }) => {
// here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // with this we prevent adding the listener if the popup is not opened
    if (!isOpen) return;
// we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const closeByEscape = (e) => {
      if (e.key === 'Escape') {
        onClose();
      }
    }

    document.addEventListener('keydown', closeByEscape)
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener('keydown', closeByEscape)
// here we watch `isOpen` to add the listener only when it’s opened
}, [isOpen, onClose])

// here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
        onClose();
    }
  }

// then we add the main wrapper with class `popup` and `popup_opened`
  return (
    <div
      className={`popup ${isOpen ? "popup_opened" : ""} popup_type_${name}`}
      onClick={handleOverlay}
    >
    {/* the container for the contents */}
      <div className='popup__container'>
        {/* here will be anything you add as `children`
*/}
        {children}
        {/* add the close button */}
        <button
          className='popup__close'
          type='button'
          onClick={onClose}
        />
      </div>
    </div>
  );
};

export default Popup;


```

And now you can use `Popup` for any popup in the project:  ` ImagePopup`, ` InfoToolTip` and `PopupWithForm`.

```jsx


function PopupWithForm({isOpen, name, onClose, ...props}) {
  return (
    <Popup isOpen={isOpen} name={name} onClose={onClose}>
        <h2 className='popup__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.

gennady-bars commented on news-explorer/src/components/SearchForm/SearchForm.js:
```diff
@@ -1,26 +1,56 @@
+            placeholder="Enter topic"
+            id='search'
+            name='newsSearch'
+            onChange={handleChange}
```
If you have `onChange` or ` onInput` in an input then you need to add attribute `value` to this input.
This is how React controls inputs.

To read more about **controlled components**: https://reactjs.org/docs/forms.html#controlled-components

gennady-bars commented on news-explorer/src/components/Main/Main.js:
```diff
@@ -7,49 +7,38 @@ import NewsCardList from '../NewsCardList/NewsCardList';
               </>} />
-          )}
-        </Routes>
+              {isLoggedIn && (
+              <Route path='/saved-news'
```
* I can visit `Saved artices` without loggin in https://disk.yandex.ru/i/ncewJi1-rih53g . But it should be protected


You need to have `ProtectedRoute` for that

gennady-bars commented on news-explorer/src/utils/protectedRoute.js:
```diff
@@ -1,10 +1,10 @@
 import React from 'react';
-import { Route, Navigate } from 'react-router-dom';
+import { Route, Navigate, Outlet } from 'react-router-dom';
 
```
`protectedRoute.js` is a component. It should be in `UpperCase` and it should be in `components`

gennady-bars reviewed (commented):
Dear Omri , (**Please, expand the general comment ↓**)

Please, scroll down the page on **github** to see my remarks right in the code

Good job but there are some remarks:


* The criteria say: When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
* When I log in I don't see the name of the user in the header in on the Saved articles page https://disk.yandex.ru/i/nEQHx5IAYwB7Qw 
* The criteria say: When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page. But when I click the **blue** icon a new black page opens
* I can't delete items from `Saved articles`  https://disk.yandex.ru/i/5foCyiyRcUqU0g . When I reload the page everything works. Please, check the logics
* I can save one and the same articles a lot of times https://disk.yandex.ru/i/JD71grzLZMpmfQ 
* I can visit `Saved artices` without loggin in https://disk.yandex.ru/i/ncewJi1-rih53g . But it should be protected.  You need to have `ProtectedRoute` for that
* `App.test.js` should be deleted because it has nothing to do with the project
* If you have `onChange` or ` onInput` in an input then you need to add attribute `value` to this input.
* `protectedRoute.js` is a component. It should be in `UpperCase` and it should be in `components`

Please, correct everything and your project will be accepted.

Good luck with the refactoring


## Number of points: 68

# Summary:
### Performance Criteria

- [] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - **When an unauthorized user clicks on the icon to save an article, the registration popup window opens.**
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - **When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.**
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [] <!--2.5--> API error handling:
  - [] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 68

## Additional comments:
-

irmo12 committed:
progress. i think

irmo12 committed:
fix save or del func,
not as DRY as id like, will improve.
next up: add check so that duplicates cannot be saved

irmo12 committed:
dev commit we'll see

irmo12 committed:
all done

irmo12 committed:
deploy

gennady-bars reviewed (commented):
Dear Omri , (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project:



* There is an error when I click **Show more**  button  https://disk.yandex.ru/i/FbkrzhvXz2VTlg 
* I can't sign in. See the error https://disk.yandex.ru/i/ECKjvImijTd18Q .

irmo12 committed:
oops

irmo12 committed:
oops

gennady-bars reviewed (approved):
Dear Omri,


You’ve done a great job.

Your project has been accepted.

Good luck

## Number of points: 92


# Summary:
### Performance Criteria

- [x] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [x] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [] <!--2.5--> API error handling:
  - [] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 92

## Additional comments:
-
